validator_client/src/key_cache.rs

```diff
@@ -0,0 +1,342 @@
+    #[serde(skip)]
+    passwords: Vec<Vec<u8>>,
+    #[serde(skip)]
+    #[serde(default = "not_decrypted")]
+    state: State,
+}
+
+type SerializedKeyMap = HashMap<Uuid, Vec<u8>>;
+
+impl KeyCache {
+    pub fn new() -> Self {
+        KeyCache {
+            uuids: Vec::new(),
+            crypto: None,
+            pairs: HashMap::new(),
+            passwords: Vec::new(),
+            state: State::DecryptedWithUnsavedUpdates,
+        }
+    }
+
+    pub fn cache_file_path<P: AsRef<Path>>(validators_dir: P) -> PathBuf {
+        validators_dir.as_ref().join(CACHE_FILENAME)
+    }
+
+    /// Open an existing file or create a new, empty one if it does not exist.
+    pub fn open_or_create<P: AsRef<Path>>(validators_dir: P) -> Result<Self, Error> {
+        let cache_path = Self::cache_file_path(validators_dir.as_ref());
+        if !cache_path.exists() {
+            Ok(Self::new())
+        } else {
+            Self::open(validators_dir)
+        }
+    }
+
+    /// Open an existing file, returning an error if the file does not exist.
+    pub fn open<P: AsRef<Path>>(validators_dir: P) -> Result<Self, Error> {
+        let cache_path = validators_dir.as_ref().join(CACHE_FILENAME);
+        let file = OpenOptions::new()
+            .write(true)
+            .read(true)
+            .create_new(false)
+            .open(&cache_path)
+            .map_err(Error::UnableToOpenFile)?;
+        serde_json::from_reader(file).map_err(Error::UnableToParseFile)
+    }
+
+    fn get_crypto_or_insert(crypto: &mut Option<Crypto>) -> &mut Crypto {
+        crypto.get_or_insert_with(|| {
+            let salt = rand::thread_rng().gen::<[u8; SALT_SIZE]>();
+            let iv = rand::thread_rng().gen::<[u8; IV_SIZE]>().to_vec().into();
+
+            let kdf = default_kdf(salt.to_vec());
+            let cipher = Cipher::Aes128Ctr(Aes128Ctr { iv });
+
+            Crypto {
+                kdf: KdfModule {
+                    function: kdf.function(),
+                    params: kdf,
+                    message: EmptyString,
+                },
+                checksum: ChecksumModule {
+                    function: Sha256Checksum::function(),
+                    params: EmptyMap,
+                    message: Vec::new().into(),
+                },
+                cipher: CipherModule {
+                    function: cipher.function(),
+                    params: cipher,
+                    message: Vec::new().into(),
+                },
+            }
+        })
+    }
+
+    fn encrypt(&mut self) -> Result<(), Error> {
+        let crypto = Self::get_crypto_or_insert(&mut self.crypto);
+
+        let secret_map: SerializedKeyMap = self
+            .pairs
+            .iter()
+            .map(|(k, v)| (*k, v.sk.serialize().as_ref().into()))
+            .collect();
+
+        let raw = bincode::serialize(&secret_map).map_err(Error::UnableToSerializeKeyMap)?;
```

Comment:
Added zeroize for `raw`. The question is: Is it ok to use `bincode::serialize` and `bincode::deserialize` if we don't know if this will copy the data at some point into a non-zeroize structure?